src/Adapter/Order/CommandHandler/AddProductToOrderHandler.php

```diff
@@ -199,22 +216,33 @@ public function handle(AddProductToOrderCommand $command)
             $orderCartRule = new OrderCartRule();
             $orderCartRule->id_order = $order->id;
             $orderCartRule->id_cart_rule = $cartRule['id_cart_rule'];
-            $orderCartRule->id_order_invoice = $invoice->id;
+            $orderCartRule->id_order_invoice = !empty($invoice->id) ? $invoice->id : 0;
             $orderCartRule->name = $cartRule['name'];
             $orderCartRule->value = $values['tax_incl'];
             $orderCartRule->value_tax_excl = $values['tax_excl'];
             $orderCartRule->add();
-
-            // @todo: use https://github.com/PrestaShop/decimal
-            $order->total_discounts += $orderCartRule->value;
-            $order->total_discounts_tax_incl += $orderCartRule->value;
-            $order->total_discounts_tax_excl += $orderCartRule->value_tax_excl;
-            $order->total_paid -= $orderCartRule->value;
-            $order->total_paid_tax_incl -= $orderCartRule->value;
-            $order->total_paid_tax_excl -= $orderCartRule->value_tax_excl;
         }
 
+        // @todo: use https://github.com/PrestaShop/decimal
+        $order->total_paid = Tools::ps_round(
+            (float) $cart->getOrderTotal(true, $totalMethod, $orderProducts),
+            $precision
+        );
+        $order->total_paid_tax_excl = Tools::ps_round(
+            (float) $cart->getOrderTotal(false, $totalMethod, $orderProducts),
+            $precision
+        );
+        $order->total_paid_tax_incl = Tools::ps_round(
+            (float) $cart->getOrderTotal(true, $totalMethod, $orderProducts),
+            $precision
+        );
+        $order->total_discounts = (float) abs($cart->getOrderTotal(true, Cart::ONLY_DISCOUNTS, $orderProducts));
+        $order->total_discounts_tax_excl = (float) abs($cart->getOrderTotal(false, Cart::ONLY_DISCOUNTS, $orderProducts));
+        $order->total_discounts_tax_incl = (float) abs($cart->getOrderTotal(true, Cart::ONLY_DISCOUNTS, $orderProducts));
+
         $order->update();
+
+        $this->contextStateManager->restoreContext();
```

Comment:
Consider restore the context if there is an exception between the moment it has been modified and the moment it is restored

try {} catch{ restore context }

Reply:
Done